javaagent-tooling/src/main/java/net/bytebuddy/agent/builder/AgentBuilderUtil.java

```diff
@@ -0,0 +1,340 @@
+      AgentBuilder.RawMatcher matcher = transformation.getMatcher();
+      // attempt to decompose the matcher and find if it applies to a named class or a subclass
+      Result result = inspect(matcher);
+      if (result == null) {
+        // we were not able to decompose the matcher
+        unoptimizedTransformations.add(transformation);
+      } else if (result.subtype) {
+        superTypeNames.addAll(result.names);
+      } else {
+        classNames.addAll(result.names);
+      }
+    }
+
+    List<?> list =
+        (List<?>)
+            Proxy.newProxyInstance(
+                AgentBuilderUtil.class.getClassLoader(),
+                new Class<?>[] {List.class},
+                (proxy, method, args) -> {
+                  String name = TransformContext.getTransformedClassName();
+                  // iterator() is the only method we expect to be called on this List
+                  if (name != null && "iterator".equals(method.getName())) {
+                    // we know that this class is going to be transformed
+                    if (classNames.contains(name) || superTypeNames.contains(name)) {
+                      return transformations.iterator();
+                    }
+                    // we already know that loading this class is going to fail, no need to
+                    // transform it
+                    if (DefineClassHandler.isFailedClass(name)) {
+                      return Collections.emptyIterator();
+                    }
+                    Set<String> loadingSuperTypes = DefineClassHandler.getSuperTypes();
+                    // super types set should contain at least java.lang.Object if this set is
+                    // empty something unexpected has happened, run all transformations
+                    if (loadingSuperTypes.isEmpty()) {
+                      return transformations.iterator();
+                    }
+                    for (String className : loadingSuperTypes) {
+                      // we know that this class is going to be transformed
+                      if (superTypeNames.contains(className)) {
+                        return transformations.iterator();
+                      }
+                    }
+
+                    // apply only the transformations that we can't decompose
+                    return unoptimizedTransformations.iterator();
+                  }
+
+                  return method.invoke(transformations, args);
+                });
+
+    agentBuilderTransformationsField.set(agentBuilder, list);
+  }
+
+  @Nullable
+  private static Result inspect(AgentBuilder.RawMatcher matcher) throws Exception {
+    if (matcher instanceof AgentBuilder.RawMatcher.Conjunction) {
+      List<AgentBuilder.RawMatcher> matchers = getDelegateMatchers(matcher);
+      if (!matchers.isEmpty()) {
+        // with our current matchers we only need to inspect the first element of the conjunction
+        return inspect(matchers.get(0));
```

Comment:
How is this going to play together with e.g. `MuzzleMatcher`, or the `not(@Decorator)` matcher? (I'm thinking that this won't break these probably)

Reply:
It won't, because muzzle and decorator matcher, which are also in the matchers list won't expand the list of possible matches. If we can tell based on the first matcher that there is no match then the result from other matchers isn't relevant. When we have a potential match then we'll let byte-buddy run all the matchers so muzzle and decorator matcher are also consulted.

Reply:
Thanks!